pkg/amiclean/ami_cleaner.go

```diff
@@ -0,0 +1,149 @@
+package amiclean
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"go.uber.org/zap"
+
+	"time"
+)
+
+const (
+	// RFC8601 is the date/time format used by AWS.
+	RFC8601 = "2006-01-02T15:04:05.000Z"
+)
+
+// AMIClean defines parameters for cleaning up AMIs based on the Branch and
+// Expiration Date.
+type AMIClean struct {
+	Delete         bool
+	Branch         string
+	Invert         bool
+	ExpirationDate time.Time
+	Logger         *zap.Logger
+	EC2Client      *ec2.EC2
+}
+
+// GetImages gets us all the private AMIs on our account so that they can be
+// looked through later. We have to do this here because the AWS API does not
+// allow you to search for AMIs by creation date or by *not* having a tag set to
+// a certain value, which would speed this up considerably.
+func (a *AMIClean) GetImages() (*ec2.DescribeImagesOutput, error) {
+	var output *ec2.DescribeImagesOutput
+
+	input := &ec2.DescribeImagesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("is-public"),
+				Values: []*string{aws.String("false")},
+			},
+		},
+	}
+
+	output, err := a.EC2Client.DescribeImages(input)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return output, err
+}
+
+// FindImagesToPurge looks through the AMIs available and produces a slice of
+// ec2.Image objects to put on the chopping block based on the contents of the
+// AMIClean struct.
+func (a *AMIClean) FindImagesToPurge(output *ec2.DescribeImagesOutput) []*ec2.Image {
+	var ImagesToPurge []*ec2.Image
+	for _, image := range output.Images {
+		imageCreationTime, _ := time.Parse(RFC8601, *image.CreationDate)
+		if imageCreationTime.After(a.ExpirationDate) {
+			continue
+		} else {
+			if a.Invert {
```

Comment:
Could you add a comment here just to make it super clear, e.g. "Select AMIs NOT tagged with branch name"?

Reply:
Definitely. I'll take care of that.

Reply:
Added some comments to this effect.